ykrt/src/compile/jitc_yk/codegen/x86_64.rs

```diff
@@ -51,15 +55,144 @@ const SYSV_CALL_STACK_ALIGN: usize = 16;
+            // after the frame containing the control point.
+            newframedst = unsafe { newframedst.byte_sub(usize::try_from(rec.size).unwrap()) };
+            if pinfo.hasfp {
+                newframedst = unsafe { newframedst.byte_add(usize::try_from(REG64_SIZE).unwrap()) };
+            }
+        }
+
+        for aotvar in rec.live_vars.iter() {
+            // Read live JIT values from the trace's stack frame.
+            let jitval = match info.lives[varidx] {
+                LocalAlloc::Stack { frame_off, size } => {
+                    let p = unsafe { jitrbp.byte_sub(frame_off) };
+                    match size {
+                        1 => unsafe { u64::from(std::ptr::read::<u8>(p as *const u8)) },
+                        2 => unsafe { u64::from(std::ptr::read::<u16>(p as *const u16)) },
+                        4 => unsafe { u64::from(std::ptr::read::<u32>(p as *const u32)) },
+                        8 => unsafe { std::ptr::read::<u64>(p as *const u64) },
+                        _ => todo!(),
+                    }
+                }
+                LocalAlloc::Register => todo!(),
+            };
+
+            let aotloc = if aotvar.len() == 1 {
+                aotvar.get(0).unwrap()
+            } else {
+                todo!("Deal with multi register locations");
+            };
+            match aotloc {
+                SMLocation::Register(reg, _size, __off, __extra) => {
+                    registers[usize::from(*reg)] = jitval;
+                }
+                SMLocation::Direct(..) => {
+                    // Due to the shadow stack we only expect direct locations for the control
+                    // point frame.
+                    debug_assert_eq!(frameid, 0);
+                    continue;
+                }
+                SMLocation::Indirect(_reg, _off, _size) => {
+                    todo!()
+                }
+                SMLocation::Constant(_v) => todo!(),
+                SMLocation::LargeConstant(_v) => todo!(),
+            }
+            varidx += 1;
+        }
+
+        // Write the return address for the previous frame into the current frame.
+        rsp = rsp.sub(REG64_SIZE);
+        ptr::write(rsp as *mut u64, rec.offset);
+    }
+
+    // Update RBP register. FIXME: We will have to do this for every frame, inside the loop above,
+    // once we can deal with multiple frames.
+    registers[usize::from(RBP_DWARF_NUM)] = frameaddr as u64;
+
+    // Write the live registers into the new stack. We put these at the very end of the new stack
+    // so that they can be immediately popped after we memcpy'd the new stack over.
+    for reg in [0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 13, 14, 15] {
+        rsp = rsp.byte_sub(REG64_SIZE);
+        ptr::write(rsp as *mut u64, registers[reg]);
+    }
+    // Now overwrite the existing stack with our newly recreated one.
+    __replace_stack(newframedst as *mut c_void, newstack, memsize);
```

Comment:
We really need to do a bit of thinking about how this interacts with stack unwinding etc. I'm pretty sure that overwriting the stack is UB *if* unwinding is possible. I don't know if it's UB if unwinding isn't possible. We should probably compile `yk` with `panic = 'abort'` but we should also think about what this means for C interpreters too. Maybe the shadow stack saves us from UB? I dunno.

Reply:
Yes we should think about this but I'm not entirely sure where to start figuring this out.

Reply:
Let's raise an issue and start putting some links in there. For example, maybe libunwind talks about some of this UB? Maybe the C spec says something about the stack?